src/Microsoft.OData.Core/UriParser/ExpressionLexer.cs

```diff
@@ -447,6 +453,24 @@ internal string AdvanceThroughBalancedParentheticalExpression()
             return expressionText;
         }
 
+        internal void TakeTextPositionSnapshot()

```

Comment:
>TakeTextPositionSnapshot [](start = 22, length = 24)

Lets have this return a bool and return false if snapshotedToken != null, indicating that the lexer only supports a single snapshot at a time.